vscode/server.ts

```diff
@@ -0,0 +1,108 @@
+import { ChildProcess, exec } from "child_process";
+import { dataform } from "df/protos/ts";
+import {
+  createConnection,
+  HandlerResult,
+  Location,
+  ProposedFeatures,
+  TextDocuments,
+  TextDocumentSyncKind
+} from "vscode-languageserver";
+import { TextDocument } from "vscode-languageserver-textdocument";
+
+const connection = createConnection(ProposedFeatures.all);
+const documents: TextDocuments<TextDocument> = new TextDocuments(TextDocument);
+let CACHED_COMPILE_GRAPH: dataform.ICompiledGraph = null;
+let WORKSPACE_ROOT_FOLDER: string = null;
+
+connection.onInitialize(() => {
+  return {
+    capabilities: {
+      textDocumentSync: TextDocumentSyncKind.Incremental,
+      // Tell the client that the server supports code completion and definitions
+      completionProvider: {
+        resolveProvider: true
+      },
+      definitionProvider: true
+    }
+  };
+});
+
+connection.onInitialized(async () => {
+  const _ = compileAndValidate();
+  const workSpaceFolders = await connection.workspace.getWorkspaceFolders();
+  // the first item is the root folder
+  WORKSPACE_ROOT_FOLDER = workSpaceFolders[0].uri;
+});
+
+connection.onRequest("compile", async () => {
+  const _ = compileAndValidate();
+});
+
+documents.onDidSave(change => {
+  const _ = compileAndValidate();
+});
+
+async function compileAndValidate() {
+  const compileResult = await getProcessResult(exec("dataform compile --json"));
```

Comment:
I don't quite understand how this works - you don't pass the project directory in, the PWD is probably not the one you want?

Reply:
I don't understand how it works either. But it does 🤷‍♂️